internal/kafka/internal/services/quota/internal/utils/cloudaccount_billing_model_resolver.go

```diff
@@ -0,0 +1,80 @@
+	v1 "github.com/openshift-online/ocm-sdk-go/accountsmgmt/v1"
+)
+
+/********************************************************************************************************************
+ * This resolver is used when the user specifies a `Cloud Account ID` in the request.
+ * The AMS Billing Model is then detected by checking what is the marketplace supported by that account.
+ ********************************************************************************************************************/
+
+var _ BillingModelResolver = &cloudAccountBillingModelResolver{}
+
+type cloudAccountBillingModelResolver struct {
+	quotaConfigProvider AMSQuotaConfigProvider
+	kafkaConfig         *config.KafkaConfig
+}
+
+func (r *cloudAccountBillingModelResolver) SupportRequest(kafka *dbapi.KafkaRequest) bool {
+	return kafka.BillingCloudAccountId != ""
+}
+
+func (r *cloudAccountBillingModelResolver) Resolve(orgId string, kafka *dbapi.KafkaRequest) (BillingModelDetails, error) {
+	kafkaBillingModel, err := r.kafkaConfig.GetBillingModelByID(kafka.InstanceType, "marketplace")
+	if err != nil {
+		return BillingModelDetails{}, err
+	}
+
+	return r.resolve(orgId, kafka, kafkaBillingModel)
+}
+
+// resolve - tries to resolve the AMS Billing Model by performing the following checks:
+// 1) Tries to retrieve the cloud account. A failure here will result in an error.
+// 2) Checks if the user has sufficient marketplace quota. If not, an error is returned.
+// 3) Tries to get the marketplace associated with the cloud provider.
+// 4) If the kafka billing model supports the marketplace detected in [3], we have found id, otherwise an error is returned.
+func (r *cloudAccountBillingModelResolver) resolve(orgId string, kafka *dbapi.KafkaRequest, kafkaBillingModel config.KafkaBillingModel) (BillingModelDetails, error) {
+	kafkaInstanceSize, err := r.kafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
+	if err != nil {
+		return BillingModelDetails{}, errors.NewWithCause(errors.ErrorGeneral, err, "error reserving quota")
+	}
+
+	quotaCosts, err := r.quotaConfigProvider.GetQuotaCostsForProduct(orgId, kafkaBillingModel.AMSResource, kafkaBillingModel.AMSProduct)
+	if err != nil {
+		return BillingModelDetails{}, errors.InsufficientQuotaError("%v: error getting quotas for product %s", err, kafkaBillingModel.AMSProduct)
+	}
+
+	idx, cloudAccount := arrays.FindFirst(getCloudAccounts(quotaCosts), func(cloudAccount *v1.CloudAccount) bool {
+		return cloudAccount.CloudAccountID() == kafka.BillingCloudAccountId &&
+			(kafka.Marketplace == "" || cloudAccount.CloudProviderID() == kafka.Marketplace)
+	})
+
+	if idx != -1 {
+		if !hasSufficientMarketplaceQuota(quotaCosts, kafkaInstanceSize.QuotaConsumed) {
+			return BillingModelDetails{}, errors.InsufficientQuotaError("quota does not support marketplace billing or has insufficient quota")
+		}
+
+		// assign marketplace in case it was not provided in the request
+		kafka.Marketplace = cloudAccount.CloudProviderID()
```

Comment:
Shouldn't this be done only when it was not provided in the request?

Reply:
This is done only if a `cloudAccount` is found (`idx != -1`) where 
```
cloudAccount.CloudAccountID() == kafka.BillingCloudAccountId &&
			(kafka.Marketplace == "" || cloudAccount.CloudProviderID() == kafka.Marketplace)
```

Reply:
Yes.

But we should respect what's provided in the request  regarding the marketplace.

If the user has provided a CloudAccountID AND a Marketplace in the API request then the marketplace value used should be the one provided in the request and not the one found in the QuotaCost response.
If it turns out that the Cloud Account ID does not exist in that Cloud Account provider then an error should be returned.

The mentioned check also checks that the Cloud Provider ID has to match the marketplace provided in the kafka request when marketplace is not empty so it is ok to unconditionally assign the returned cloudprovider id in the quotacost. It is unnecessary but it would be correct and it avoids using one additional conditional in the code :+1: